### ISSUE TITLE
Allow previously-synced jobs to run without network access to source control

### ISSUE THREAD

**Issue by AlanCoding** [ansible/awx#287]
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
 - API

##### SUMMARY

Network access to source control should not be required for all jobs, since the project folder already exists with the correct revision.

Scenario: User has a SCM server where the connection is flaky / goes down for maintenance / etc.

##### STEPS TO REPRODUCE

Start using a single-instance deployment of AWX

 - Run a normal successful Job Template which runs against localhost
 - Disconnect your internet
 - Run that JT again

##### EXPECTED RESULTS

Job is fully successful, because nothing involved requires network access.

##### ACTUAL RESULTS

Job is not successful, because the dependent project sync can not reach out to source control

##### ADDITIONAL INFORMATION

I have a start on this here:

https://github.com/ansible/awx/compare/devel...AlanCoding:git_lazy?expand=1

Would love to get community feedback, or welcome anyone who wants to get involved in contributing.

It would also be nice to consolidate the project update playbook into roles so that the output becomes cleaner.

**Comment by wenottingham** [comment]
A really cruddy way to do this for a subset of cases:
```
diff --git a/awx/playbooks/project_update.yml b/awx/playbooks/project_update.yml
index 20c84b666..ca9712b13 100644
--- a/awx/playbooks/project_update.yml
+++ b/awx/playbooks/project_update.yml
@@ -17,6 +17,16 @@
   gather_facts: false
   tasks:
 
+    - name: read existing project version
+      command: git rev-parse HEAD
+      args:
+         chdir: "{{ project_path }}"
+      register: existing_version
+
+    - name: see if version is already checked out
+      meta: end_play
+      when: scm_full_checkout|default('') and existing_version['stdout'] == scm_branch and not scm_delete_on_update|default('') and not scm_clean|default('')
+
     - name: delete project directory before update
       file:
         path: "{{project_path|quote}}"
```

Would need to expand this to other SCMs, figure out where/how to handle roles in this case, and potentially enhance it to work when 'clean' is set on repos that are already clean (via `git status`).

**Comment by wenottingham** [comment]
... and yes, to do this "properly" will likely involve reorganizing the playbook.

**Comment by AlanCoding** [comment]
That looks pretty similar in function to [this branch](https://github.com/ansible/awx/compare/devel...AlanCoding:git_lazy?expand=1), although not the same in execution. It's an open question about whether there would be value in going through the git module vs running a command, since there's kind of a tradeoff between clarity and using the "standard" tool.

> Would need to expand this to other SCMs

Yes, @rooftopcellist did that exact thing, and his report as I remember was that it works following the exact same pattern with Mercurial. This might be a good argument for sticking with the modules as a means of reading the revision by a command.

As long as we're agreed on the particular style, the remaining work is yak shaving (in addition to subversion, if anyone cares). I agree it would also be nice to consolidate into roles to reduce the noise in a project update.

**Comment by wenottingham** [comment]
Yeah, I prototyped with the commandline b/c it would be nice to pull the revision & 'clean' status at once. That may not be practical.

**Comment by AlanCoding** [comment]
good catch, I had not taken into account the 'clean' status.

**Comment by perzizzle** [comment]
This would be very useful for our environment. We had an entire datacenter go down that contained our git repository. We use tower to recover from these kinds of outages but we couldn't use tower until we recovered git increasing the total time of outage. Obviously we should have had a better git setup as well...

**Comment by wenottingham** [comment]
Closed per #567 (*for git only*)

**Comment by Testicule** [comment]
"Closed per #567 (for git only)"

git only. not really good enough.

I fail to see why there is any need to contact source control when the project is checked out and a job can be run.
Why?
Why is it necessary?

Launch a job:
 - If the project has "Update revision on launch" checked - then contact SCM.
 - If you don't have the project checked out - just fail, or, contact SCM, I don't care.
 - If you've got the project checked out, then run the job.

Why is this so difficult?
